src/persist-client/src/internal/compact.rs

```diff
@@ -583,17 +593,16 @@ where
             if remaining_updates_by_run[index] == 0 {
                 // repopulate from the originating run, if any parts remain
                 let (part_desc, parts) = &mut runs[index];
-                if let Some(part) = parts.next() {
+                if let Some(part) = parts.pop_front() {
                     let start = Instant::now();
-                    let mut part = fetch_batch_part(
-                        shard_id,
-                        blob.as_ref(),
-                        &metrics,
-                        &metrics.read.compaction,
-                        &part.key,
-                        part_desc,
-                    )
-                    .await?;
+                    let mut part = part
+                        .join(shard_id, blob.as_ref(), &metrics, part_desc)
+                        .await?;
+                    // start_prefetches is O(n) so calling it here is O(n^2). N
+                    // is the number of things we're about to fetch over the
+                    // network, so if it's big enough for N^2 to matter, we've
+                    // got bigger problems.
```

Comment:
idk if it's even worth a comment here, but we could mention it IS possible to make this linear if we do the bookkeeping of which part we last considered / how much budget we have / how large the last part we popped off was, but it's not worth the effort+safety concerns for now